monorepo: Merge `kvm-bindings` into `kvm-ioctls`

TimePrinciple opened the pull request:
### Summary of the PR

Fixes #291 

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR have Signed-Off-By trailers (with
  `git commit -s`), and the commit message has max 60 characters for the
  summary and max 75 characters for each description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [ ] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

andreeaflorescu committed:
first commit - copied code from kvm_wrapper

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
added travis config file

The CI script is currently running the release build with the default
features and with kvm-v4_14_0 and kvm-v4_20_0 features. Similarly
cargo test is also run with the available features

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
added travis build status badge to README.md

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
added crate version & license badges to readme

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
do not enforce rust Edition 2018

This crate should work with older rust as well.

Fixes: https://github.com/rust-vmm/kvm-bindings/issues/1

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
update to 0.1.1 version

Signed-off-by: Andreea Florescu <fandree@amazon.com>

Commit:
CODEOWNERS: Make PullAssigner the default owner

The PullAssigner bot [1] will pick 2 github handles from the gate
keepers list [2] and assign them for reviewing any new PR.
The bot goes uses a round robin algorithm to select handles from the
list.

What happens when the CODEOWNERS file gets populated with real
owners? Following the PR review and approval process [3], the
PullAssigner bot will either pick only one or zero reviewers from the
gate keepers list, if the CODEOWNERS file points to exactly one or more
than one actual ownwer, respectively.

[1] https://pullpanda.com/assigner
[2] https://github.com/rust-vmm/community/blob/master/GATEKEEPERS.md
[3] https://github.com/rust-vmm/community#pr-review-and-approval

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

acatangiu committed:
implement CpuId as a FamStructWrapper over kvm_cpuid2

Since kvm_cpuid2 contains a flexible array member, we can implement
CpuId as a FamStructWrapper<kvm_cpuid2>. This allows users of
kvm-ioctls to work with safe code even when dealing with kvm_cpuid2.

Signed-off-by: Adrian Catangiu <acatan@amazon.com>

acatangiu committed:
implement Msrs as a FamStructWrapper over kvm_msrs

Since kvm_msrs contains a flexible array member, we can implement
Msrs as a FamStructWrapper<kvm_msrs>. This allows users of kvm-ioctls
to work with safe code even when dealing with kvm_msrs structures.

Signed-off-by: Adrian Catangiu <acatan@amazon.com>

acatangiu committed:
implement MsrList as a FamStructWrapper over kvm_msr_list

Since kvm_msr_list contains a flexible array member, we can implement
MsrList as a FamStructWrapper<kvm_msr_list>. This allows users of
kvm-ioctls to work with safe code even when dealing with kvm_msr_list.

Signed-off-by: Adrian Catangiu <acatan@amazon.com>

acatangiu committed:
use a 'fam-wrappers' feature to gate the FamStructWrappers

The FamStructWrappers definitions as well as the vmm-sys-util
dependency are now gated by an opt-in `fam-wrappers` feature.

Signed-off-by: Adrian Catangiu <acatan@amazon.com>

acatangiu committed:
remove git tracking for Cargo.lock

Also add Cargo.lock to .gitignore

Signed-off-by: Adrian Catangiu <acatan@amazon.com>

acatangiu committed:
bump version to 0.1.2

Signed-off-by: Adrian Catangiu <acatan@amazon.com>

acatangiu committed:
update bump from v0.1.2 to v0.2.0

There was newly added functionality which warants a version number
`minor` bump, not just a version number `patch` bump.

Signed-off-by: Adrian Catangiu <acatan@amazon.com>

andreeaflorescu committed:
Switch to rust-vmm-ci

Use Buildkite and rust-vmm-ci for testing.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
fam_wrappers: simplify imports

Import bindings from x86::bindings so we don't need to have the
same logic to exclude features twice (once in the x86/mod.rs and
once in x86/fam_wrappers.rs).

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
make kernel bindings mutually exclusive

When the crate is built with --all-features we have a bunch of errors
because we are not excluding the 4.14 and 4.20 bindings. Whenever
all features is specified now, the 4.20 bindings are going to be
used. This change is required for switching to rust-vmm-ci where
the tests are run using --all-features flag.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
rust-vmm-ci: disable clippy warnings

Autogenerated code throws lots of clippy errors. Disable the
warnings as we don't want to manually edit autogenerated code.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
fix link issues on aarch64 musl

The linker was unable to find __addtf3, __multf3 and __subtf3.

Added target-feature=+crt-static and link-arg=-lgcc as a temporary
workaround. This seems to be the accepted fix in the Rust community:
rust-lang-nursery/compiler-builtins#201

A permanent fix is yet to be implemented in the Rust compiler.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
added tests for missing combination of features

We have a few combination of features that were not tested with the CI
due to switching to rust-vmm-ci and adding the fam-wrappers feature.

Added a new Buildkite pipeline for the missing tests.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
removed travis config

We are now using rust-vmm-ci for testing.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

dianpopa committed:
arm: add wrapper over `kvm_reg_list`

In order to be able to save/restore vcpu registers
on arm, we need to support the KVM_REG_LIST
ioctl. However, the structure holding the registers
list has a flexible array member.
This commit add a `FamStructWrapper` over the `kvm_reg_list`
structure.

Fixes: https://github.com/rust-vmm/kvm-bindings/issues/20.

Signed-off-by: Diana Popa <dpopa@amazon.com>

dianpopa committed:
arm: adjust buildkite pipeline

Enable build with the v4.14 bindings and `fam-wrappers`
feature on arm.

Signed-off-by: Diana Popa <dpopa@amazon.com>

dianpopa committed:
bump version to 0.3.0

The CHANGELOG was also updated with newly
added functionality.

Signed-off-by: Diana Popa <dpopa@amazon.com>

alexandruag committed:
update rust-vmm-ci anchor

This update brings the following commits:

e58ea74 Fix kcov_ouput_dir typo in test_coverage.py
d62d781 fix buildkite typos in readme
0fc8ced refactor test_benchmark.py
741b894 checkout to PR branch before finishing test_bench
645a5c3 test_bench: don't crash when no bench on master
bd32544 Fetch origin in benchmark test
35beb91 Fix commit message test
53427aa benchmarks: add test that can run at every PR
abd2c90 Add test for commit message format
fe859f4 Update container image to v6
75d7254 run cargo check on all features
7e3f307 skip coverage-arm test
cd7096e Enable rust-vmm coverage test in CI

Signed-off-by: Alexandru Agache <aagch@amazon.com>

alexandruag committed:
rename coverage_config and adjust coverage target

The path and threshold have been updated due to changes in the CI.
Added a coverage config file for aarch64 as well (coverage testing
on aarch64 is disabled for now because of tooling issues).

Signed-off-by: Alexandru Agache <aagch@amazon.com>

lauralt committed:
document potential undefined behavior

Mentioned in README about potential undefined behavior with
incompatible versions.

Signed-off-by: Alexandru Agache <aagch@amazon.com>

alishir committed:
Update verstion to the latest release.

Signed-off-by: Ali Shirvani <aj.shirvani@gmail.com>

rbradford committed:
Release 0.4.0

This release is to allow kvm-ioctls to be released with an increased
dependency on kvm-bindings rather than pinning to an older version.
Update the vmm-sys-utils version to reflect the latest version and to
match the one from kvm-ioctls.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

dependabot-preview[bot] committed:
Bump rust-vmm-ci from `e58ea74` to `24d66cd`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci)
from `e58ea74` to `24d66cd`.

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

lauralt committed:
update coverage

Signed-off-by: Laura Loghin <lauralg@amazon.com>

chenhengqi committed:
fix typo

Signed-off-by: Hengqi Chen <chenhengqi@outlook.com>

dependabot-preview[bot] committed:
Upgrade to GitHub-native Dependabot

dependabot[bot] committed:
Bump rust-vmm-ci from `24d66cd` to `d2ab3c0`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `24d66cd` to `d2ab3c0`.
- [Release notes](https://github.com/rust-vmm/rust-vmm-ci/releases)
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/24d66cdae63d4aa7f8de01b616c015b97604a116...d2ab3c090833aec72eee7da1e3884032206b00e3)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

andreeaflorescu committed:
update coverage

Coverage update was caused by upgrading to Rust 1.52.1.

Since this is a significant change in coverage, I took a look at the
difference in the coverage report from the previous Rust version and
this one.

Previously, kcov reported 3192 instrumented lines, and 1780 executed
lines.
With the new version, kcov reports 2608 instrumented lines, and 1781
executed lines.

Looking at the coverage report I can't notice anything fishy.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

Commit:
arm, arm64, x86: Remove v4.14 bindings

This patch makes the bindings v4.20 the only available version for all
architectures. Because of this unique possibility, the code is
simplified, removing the specific Rust features to pick a specific
bindings version.

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

Commit:
arm64, x86: Update bindings to v5.13

Replace the v4.20 bindings with the up to date v5.13 ones. The `arm`
architecture is now covered with the `arm64` bindings the same way both
`x86` and `x86_64` are covered by the bindings from `x86_64`.

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `d2ab3c0` to `7693628`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `d2ab3c0` to `7693628`.
- [Release notes](https://github.com/rust-vmm/rust-vmm-ci/releases)
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/d2ab3c090833aec72eee7da1e3884032206b00e3...769362803545f1d8abcffba77a9beca0bca8953e)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `7693628` to `877d1fb`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `7693628` to `877d1fb`.
- [Release notes](https://github.com/rust-vmm/rust-vmm-ci/releases)
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/769362803545f1d8abcffba77a9beca0bca8953e...877d1fb680dcebb33124e6953969722b483fca8f)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

lauralt committed:
Release 0.5.0

In this release, the v4.14 bindings were removed, and the v4.20
ones were replaced with v5.13.

Signed-off-by: Laura Loghin <lauralg@amazon.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `877d1fb` to `8901e77`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `877d1fb` to `8901e77`.
- [Release notes](https://github.com/rust-vmm/rust-vmm-ci/releases)
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/877d1fb680dcebb33124e6953969722b483fca8f...8901e7752288ae1061e2ee888a104c083a451668)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `8901e77` to `1311bfa`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `8901e77` to `1311bfa`.
- [Release notes](https://github.com/rust-vmm/rust-vmm-ci/releases)
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/8901e7752288ae1061e2ee888a104c083a451668...1311bfa03f3510a3aec6b69ae085f21e8e4a1e5e)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

catdum committed:
Deprecate custom pipeline

We can autogenerate the custom pipelines as well now.

Signed-off-by: Catalin Dumitru <catdum@amazon.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `1311bfa` to `ae7db2d`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `1311bfa` to `ae7db2d`.
- [Release notes](https://github.com/rust-vmm/rust-vmm-ci/releases)
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/1311bfa03f3510a3aec6b69ae085f21e8e4a1e5e...ae7db2d98a071f52de3d60af9c937204b1f087a4)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

andreeaflorescu committed:
update dependabot formatting

Signed-off-by: Andreea Florescu <fandree@amazon.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `ae7db2d` to `f67ef4c`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `ae7db2d` to `f67ef4c`.
- [Release notes](https://github.com/rust-vmm/rust-vmm-ci/releases)
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/ae7db2d98a071f52de3d60af9c937204b1f087a4...f67ef4c84d99facfdba70a97eb11a4364267eb2b)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `f67ef4c` to `68d4dbf`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `f67ef4c` to `68d4dbf`.
- [Release notes](https://github.com/rust-vmm/rust-vmm-ci/releases)
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/f67ef4c84d99facfdba70a97eb11a4364267eb2b...68d4dbf85c2ef54a440afa635d2f0db8635dcfb7)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `68d4dbf` to `7931077`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `68d4dbf` to `7931077`.
- [Release notes](https://github.com/rust-vmm/rust-vmm-ci/releases)
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/68d4dbf85c2ef54a440afa635d2f0db8635dcfb7...7931077cdc577edc20af54bf4786de750886fb85)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `7931077` to `b037be3`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `7931077` to `b037be3`.
- [Release notes](https://github.com/rust-vmm/rust-vmm-ci/releases)
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/7931077cdc577edc20af54bf4786de750886fb85...b037be339677c2f24b7ba676fc9ff893ad474305)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

andreeaflorescu committed:
update dependabot frequency

Signed-off-by: Andreea Florescu <fandree@amazon.com>

wllenyj committed:
fam_wrappers: implement PartialEq

Implementing PartialEq as helpful for testing. Since the
FamStructWrapper type implementation of PartialEq requires the type
T to implement PartialEq, it is necessary here to compare the
other fields of the type T except entries. entries will be compare
within FamStructWrapper's PartialEq.

Signed-off-by: wllenyj <wllenyj@linux.alibaba.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `b037be3` to `d216a46`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `b037be3` to `d216a46`.
- [Release notes](https://github.com/rust-vmm/rust-vmm-ci/releases)
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/b037be339677c2f24b7ba676fc9ff893ad474305...d216a46879096a4f1e83af3a7b6c90c26e1a688d)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

Commit:
Fix warnings and adjust test coverage

Signed-off-by: Sergii Glushchenko <gsserge@amazon.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `d216a46` to `99fe2eb`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `d216a46` to `99fe2eb`.
- [Release notes](https://github.com/rust-vmm/rust-vmm-ci/releases)
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/d216a46879096a4f1e83af3a7b6c90c26e1a688d...99fe2eb2e05d1b2cbeed6fb00b754e8f1c5b2f81)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `99fe2eb` to `258161e`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `99fe2eb` to `258161e`.
- [Release notes](https://github.com/rust-vmm/rust-vmm-ci/releases)
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/99fe2eb2e05d1b2cbeed6fb00b754e8f1c5b2f81...258161e88af86e1fb1fb188de25fd0e1f9d26d0a)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `258161e` to `5f36cc9`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `258161e` to `5f36cc9`.
- [Release notes](https://github.com/rust-vmm/rust-vmm-ci/releases)
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/258161e88af86e1fb1fb188de25fd0e1f9d26d0a...5f36cc9604fbe8beb2dc9cc27d2a551300b741a5)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `5f36cc9` to `7af2e16`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `5f36cc9` to `7af2e16`.
- [Release notes](https://github.com/rust-vmm/rust-vmm-ci/releases)
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/5f36cc9604fbe8beb2dc9cc27d2a551300b741a5...7af2e16d9150869f252358a66adb2c5da1deec75)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

andreeaflorescu committed:
allow undocumented unsafe blocks

This is autogenerated code so it's not feasible to add manual
documentation for it.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

studychao committed:
versions: update vmm-sys-util version

In order to use caret dependency, we should specify vmm-sys-util
dependency to the latest version (0.11.0)

fixes: #74

Signed-off-by: Chao Wu <chaowu@linux.alibaba.com>

roypat committed:
release v0.6.0

Details in CHANGELOG.md

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

dependabot[bot] committed:
Bump rust-vmm-ci from `7af2e16` to `aa33f19`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `7af2e16` to `aa33f19`.
- [Release notes](https://github.com/rust-vmm/rust-vmm-ci/releases)
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/7af2e16d9150869f252358a66adb2c5da1deec75...aa33f19d6bf4b5eef838d127b09c28900cafbdd6)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `aa33f19` to `607c775`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `aa33f19` to `607c775`.
- [Release notes](https://github.com/rust-vmm/rust-vmm-ci/releases)
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/aa33f19d6bf4b5eef838d127b09c28900cafbdd6...607c775ddec0d77a2865520d3f9f7c211d9fb7c7)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `607c775` to `c2f8c93`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `607c775` to `c2f8c93`.
- [Release notes](https://github.com/rust-vmm/rust-vmm-ci/releases)
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/607c775ddec0d77a2865520d3f9f7c211d9fb7c7...c2f8c93e3796d8b3ea7dc339fad211457be9c238)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

rbradford committed:
docs: Updating CONTRIBUTING.md

* Update for newer bindgen
* Update to reflect that there are not multiple versions of bindings any
  more

Signed-off-by: Rob Bradford <rbradford@rivosinc.com>

rbradford committed:
arm64, x64: Update bindings for the existing architectures

This is a significant change as the kernel headers have changed
significantly and bindgen has changed some names (and some names have
changed due to file ordering changes.)

Depending on which structures are used this change may result in an API
or ABI change.

Signed-off-by: Rob Bradford <rbradford@rivosinc.com>

rbradford committed:
x86_64: Rename from x86 for consistency

This will break the API but makes it consistent with different
architectures (so it matches the kernel arch name that it was generated
from).

Signed-off-by: Rob Bradford <rbradford@rivosinc.com>

rbradford committed:
misc: Update CHANGELOG for API changes

Signed-off-by: Rob Bradford <rbradford@rivosinc.com>

rbradford committed:
misc: Drop x86 (32-bit support)

Signed-off-by: Rob Bradford <rbradford@rivosinc.com>

likebreath committed:
docs: Fix a wrong path from CONTRIBUTING.md

Signed-off-by: Bo Chen <chen.bo@intel.com>

Commit:
chore: Remove broken `build status`

Signed-off-by: Jonathan Woollett-Light <jcawl@amazon.co.uk>

dependabot[bot] committed:
Bump rust-vmm-ci from `c2f8c93` to `8627b37`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `c2f8c93` to `8627b37`.
- [Release notes](https://github.com/rust-vmm/rust-vmm-ci/releases)
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/c2f8c93e3796d8b3ea7dc339fad211457be9c238...8627b3766b2bedde4657c7e9ddfc6f95a20e6942)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `8627b37` to `7c1057e`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `8627b37` to `7c1057e`.
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/8627b3766b2bedde4657c7e9ddfc6f95a20e6942...7c1057e9bcba7ed5090fd62b9fef0570fadb06e6)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `7c1057e` to `be28ad8`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `7c1057e` to `be28ad8`.
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/7c1057e9bcba7ed5090fd62b9fef0570fadb06e6...be28ad8e2c248875637d791c40282ec3d6e9c639)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

00xc committed:
Build docs.rs documentation with features enabled

The current documentation build in docs.rs does not include any
crate features (at the moment just "fam-wrappers"), which might lead
to users of the crate not being aware of their existence.

Signed-off-by: Carlos López <carlos.lopez@suse.com>

roypat committed:
Exclude autogenerated bindings from coverage report

bindgen not putting its tests inside of test modules confuses the tool,
resulting in a coverage drop of ~30%.

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

dependabot[bot] committed:
Bump rust-vmm-ci from `be28ad8` to `0100de0`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `be28ad8` to `0100de0`.
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/be28ad8e2c248875637d791c40282ec3d6e9c639...0100de0f57ec8d9f55141b43b7332a5d7f357496)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

roypat committed:
Update vmm-sys-util to 0.12.1

The update brings a fix for a security vulnerability behind
feature-gated code not used by kvm-bindings (the `with-serde` feature),
see GHSA-875g-mfp6-g7f9

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

roypat committed:
Prepare release 0.7.0

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

likebreath committed:
Remove unnecessary clippy override

The clippy override `allow(deref_nullptr)` is no longer needed since the
issue has been fixed from rust-bindgen [1].

[1] https://github.com/rust-lang/rust-bindgen/issues/1651

Signed-off-by: Bo Chen <chen.bo@intel.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `0100de0` to `72ebeb3`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `0100de0` to `72ebeb3`.
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/0100de0f57ec8d9f55141b43b7332a5d7f357496...72ebeb393cf0843fef7b2b543058883095780797)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump rust-vmm-ci from `72ebeb3` to `7606478`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `72ebeb3` to `7606478`.
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/72ebeb393cf0843fef7b2b543058883095780797...7606478202667234d9378d6850bfda9c219eee7b)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

roypat committed:
Add `serde` support for various bindings

Add implementations of `serde::{Serialize, Deserialize}` for
kvm-bindings structures commonly used for live migration/snapshotting.
The selections of structures to add the derives to was determined by
finding the minimal set that allows both Firecracker and CloudHypervisor
to compile against this branch.

Serialization happens as opaque binary blobs via the `zerocopy` crate.
This is to support live migration across different versions of
kvm-bindings (which can potentially be using bindings generated from
different versions of the kernel headers). Since the kernel's ABI is
backward-compatible, the only changes that can happen are:

- "reserved" areas of structures acquire meaning. This is a change
  purely at the representational level, and has no impact on the
  underlying binary representation. Thus, our serialization is
  roboust to these changes (whereas simply deriving the serde traits
  would cause incompatibilities here).

- Adding new fields to the end of a structure. When deserializating, we
  compare the length of the deserialized data with the size of the
  struct, and pad with zeros/truncate as appropriate.

This allows seamless live update.

We pick the `zerocopy` crate for binary serialization as it is already
in use in the rust-vmm ecosystem (in `acpi_tables`), and it has a very
strong security stance (for example, using formal verification in their
CI). The alternative would have been using `ByteValued` from
`vm-memory`, but I do not want `kvm-bindings` to gain a dependency on
`vm-memory` (especially since `vm-memory` brings in a lot more
functionality that `kvm-bindings` does not need).

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

roypat committed:
test: test with-serde feature in CI

Do a build-test of the `with-serde` feature. Also do some simple
round-trip testing to ensure consistent serialization (particularly, for
the items that have manual (De)Serialize implementations).

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

roypat committed:
Support treating kvm_xsave as a FamStruct

In linux 5.17, kvm_xsave got turned into a FamStruct by adding the
flexible "extra" member to its definition. However, unlike all other
such structs, it does not contain a "length" field. Instead, the length
of the flexible array member has to be determined by querying the
`KVM_CAP_XSAVE2` capability. This requires access to a VM file
descriptor, and thus cannot happen in the `FamStruct::len` trait method.
To work around this, define a wrapper struct that caches the length of a
previous `KVM_CHECK_EXTENSION(KVM_CAP_XSAVE2)` call, and implement
`FamStruct` for this wrapper. Then in kvm-ioctls, we can expose a
function that first query `KVM_CAP_XSAVE2`, then invokes
`KVM_GET_XSAVE2` to retrives the `kvm_xsave` structure, and then combine
them into the below `kvm_xsave2` structure to be managed as a
`FamStruct`.

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

roypat committed:
doc: Add instructions for generating bindings

Describe what needs to be done to not break serialization support.

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

roypat committed:
fix: Remove "check-warnings" step from custom ci steps

We already have a `cargo check --all-targets --all-features` with `-D
warnings` in the default CI inherited from rust-vmm-ci, so remove this
duplicate.

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

roypat committed:
doc: Add reference to CONTRIBUTING.md to README

The contributing file contains instructions for how to (re)generate
bindings, but this is a somewhat unintuitive location for these
instructions. Thus explicitly refer to it from the README.

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

roypat committed:
doc: Add CHANGELOG entry

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

roypat committed:
Prepare 0.8.0 release

This release contains the optional serialization support via serde.

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

roypat committed:
fix: derive `Default` for `kvm_xsave2`

`FamStructWrapper<T>` imposes a bound of `Default + FamStruct` on `T`.
However, it seems that when defining type aliases, Rust will not ensure
this bound is met until the type alias is actually used (as in a
function definition). This made the `Xsave` type defined in
`x86_64/fam.rs` essentially "unmentionable", as any attempts as using it
will result in a compile error.

Fix this by deriving `Default` for it, and also derive `Debug` for
consistency with other types.

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

roypat committed:
doc: document `kvm_xsave2::len` field

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

roypat committed:
test: Ensure `Xsave` FAMWrapper can be used

Regression test for missing `Default` implementation which was causing
compile errors when downstream uses referred to `Xsave` structure.

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

roypat committed:
Prepare 0.8.1 release

This release contains a hotfix for a bug that made the `Xsave`
FAMStructWrapper unusable in downstream code.

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

roypat committed:
chore: Update coverage

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

roypat committed:
Avoid temporary Vec<u8> allocation during deserialization

Instead of reusing `Vec::<u8>::deserialize` to defer the
deserialization, implement a Visitor that can construct `$typ` directly
from the deserialization byte stream. This eliminates a temporary
allocation where the input byte stream is first copied into a `Vec<u8>`
and later copied to the stack to construct `$typ` via `transmute!`.

This saves a few microseconds (up to 1ms on specific hardware) when
deserializing microVM snapshots.

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

roypat committed:
chore: Prepare 0.8.2 release

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

RuoqingHe committed:
arm64, x86_64: Update bindings from v6.9 kernel source

Update existing architecture's bindings.

Signed-off-by: Ruoqing He <ruoqing@isrc.iscas.ac.cn>

MatiasVara committed:
Prepare 0.9.0 release

Signed-off-by: Matias Ezequiel Vara Larsen <mvaralar@redhat.com>

dependabot[bot] committed:
build(deps): bump rust-vmm-ci from `7606478` to `9f641f2`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `7606478` to `9f641f2`.
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/7606478202667234d9378d6850bfda9c219eee7b...9f641f269b987aeec4a771f459f6d2af28cf7288)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

likebreath committed:
Test manual (De)Serialize implementations with serde_json

The 'serde_json' crate has been commonly used to (de)serialize Rust data
structures in the format of JSON, such as Cloud Hyervisor uses this
crate for its live migration support. This patch adds a set of unit
tests to ensure our manual (De)Serialize implementations works with
the 'serde_json' crate.

Signed-off-by: Bo Chen <chen.bo@intel.com>

likebreath committed:
Fix manual (De)Serialize implementations with serde_json

The new Deserializer implementation (#107) does not work with the
'serde_json' crate, particularly `serde_json::{to_string,
from_str}`. The new implementation always expects `byte array` for
deserilaiztion. Meanwhile, the 'serde_json' crate always (de)serializes
`String` as `Vector` [1][2] which is treated as `sequence` (e.g. not
'byte array') by default for (de)serialization [3].

To fix this issue, this patch extends the new Deserializer
implementation to support both `byte array` and `sequence`.

[1] https://github.com/serde-rs/json/blob/cc7a1608c9bb7736c884926e016421af41a1ebe7/src/ser.rs#L2168-L2175
[2] https://github.com/serde-rs/json/blob/cc7a1608c9bb7736c884926e016421af41a1ebe7/src/de.rs#L30-L39
[3] https://serde.rs/data-model.html#types

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>
Signed-off-by: Bo Chen <chen.bo@intel.com>

RuoqingHe committed:
chore: Prepare 0.9.1 release

Signed-off-by: Ruoqing He <heruoqing@iscas.ac.cn>

RuoqingHe committed:
riscv64: bump rust-vmm-ci from `9f641f2` to `55ee075`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `9f641f2` to `55ee075`.
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/9f641f2...55ee075)

Which has RISC-V CI ready to be enabled.

Signed-off-by: Ruoqing He <heruoqing@iscas.ac.cn>

RuoqingHe committed:
riscv64: Introduce RISC-V bindings

Introduce RISC-V KVM bindings for Linux kernel v6.9, and enable RISC-V
CI in `.platform` file.

Signed-off-by: Ruoqing He <heruoqing@iscas.ac.cn>

RuoqingHe committed:
arm64: Fix unexpected architecture warning

Replace `aarch` with `arm` to meet the expected architecture.

Signed-off-by: Ruoqing He <heruoqing@iscas.ac.cn>

RuoqingHe committed:
x86_64: Fix code coverage

Setting code coverage to 79.17% as `test_coverage.py` reported.

Signed-off-by: Ruoqing He <heruoqing@iscas.ac.cn>

RuoqingHe committed:
riscv64: Introduce `fam-wrappers` feature

Implement `fam-wrappers` to get hold on `kvm_reg_list` over RISC-V
platforms.

Signed-off-by: Ruoqing He <heruoqing@iscas.ac.cn>

RuoqingHe committed:
riscv64: Introduce `serde` feature

Implement `serde` for basic RISC-V register structures to help VM
snapshotting.

Signed-off-by: Ruoqing He <heruoqing@iscas.ac.cn>

RuoqingHe committed:
riscv64: Enable custom-tests

Enable custom build-gnu tests for RISC-V.

Signed-off-by: Ruoqing He <heruoqing@iscas.ac.cn>

dependabot[bot] committed:
build(deps): bump rust-vmm-ci from `55ee075` to `438f36e`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `55ee075` to `438f36e`.
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/55ee075f855fd62c3aa36d3fc62e74a95b5b74b4...438f36e242c4655f0d097f96001ad7f852c44bf5)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

RuoqingHe committed:
chore: Prepare 0.10.0 release

Introduce RISC-V KVM bindings.

Signed-off-by: Ruoqing He <heruoqing@iscas.ac.cn>

dependabot[bot] committed:
build(deps): bump rust-vmm-ci from `438f36e` to `cdb4a2d`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `438f36e` to `cdb4a2d`.
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/438f36e242c4655f0d097f96001ad7f852c44bf5...cdb4a2d5cd207846b51407c96a01a175efc0b274)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

RuoqingHe committed:
chore: Update `bindgen` used to latest

The 0.70.1 version of `bindgen` provides the same definition and better
readability of tests generated.

Improve generated code readability while leaving its functionality
intact and unchanged.

Signed-off-by: Ruoqing He <heruoqing@iscas.ac.cn>

RuoqingHe committed:
chore: Update `binding.rs`

Update generated bindings of `x86_64`, `arm64` and `riscv64` to apply
`bindgen` upgrade to version 0.70.1.

Signed-off-by: Ruoqing He <heruoqing@iscas.ac.cn>

RuoqingHe committed:
monorepo: Move kvm-ioctls related files to subdirectory

This change is needed to be able to merge the kvm-bindings crate into
this repository and prevent conflicts from happening.

Signed-off-by: Ruoqing He <heruoqing@iscas.ac.cn>

RuoqingHe committed:
monorepo: Finalize repository tree layout

Reorganize files from `kvm-bindings` and `kvm-ioctls`, and layout them
like a monorepo.

Signed-off-by: Ruoqing He <heruoqing@iscas.ac.cn>

RuoqingHe committed:
monorepo: Setup repository as a Rust workspace

Create Rust workspace to incorporate `kvm-bindings` and `kvm-ioctls` as
members of this workspace.

Signed-off-by: Ruoqing He <heruoqing@iscas.ac.cn>

dependabot[bot] committed:
build(deps): bump rust-vmm-ci from `cdb4a2d` to `1150c47`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `cdb4a2d` to `1150c47`.
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/cdb4a2d5cd207846b51407c96a01a175efc0b274...1150c47872d4a1376cf5d95e15af7e45cd6d27b9)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

roypat commented:
@rbradford are you okay with this approach as well? e.g. merging kvm-bindings into the kvm-ioctls repository, and renaming this repo to just "kvm", instead of moving both ioctls and bindings to a new repository (the idea being that this way, we preserve stars and other properties of this repository)

TimePrinciple commented:
> @rbradford are you okay with this approach as well? e.g. merging kvm-bindings into the kvm-ioctls repository, and renaming this repo to just "kvm", instead of moving both ioctls and bindings to a new repository (the idea being that this way, we preserve stars and other properties of this repository)

I remember @rbradford  is on vacation and will be back on 4th Nov 😂 (if the date he left on his slack is correct)

rbradford commented:
> @rbradford are you okay with this approach as well? e.g. merging kvm-bindings into the kvm-ioctls repository, and renaming this repo to just "kvm", instead of moving both ioctls and bindings to a new repository (the idea being that this way, we preserve stars and other properties of this repository)

Sounds good to me!

roypat commented on .buildkite/custom-tests.json:
```diff
@@ -0,0 +1,38 @@
+{
+    "tests": [
+        {
+            "test_name": "build-fam-gnu",
+            "command": "cargo build --release --features=fam-wrappers",
+            "platform": [
+                "x86_64",
+                "aarch64",
+                "riscv64"
+            ]
+        },
+        {
+            "test_name": "build-fam-musl",
+            "command": "cargo build --release --features=fam-wrappers --target {target_platform}-unknown-linux-musl",
+            "platform": [
+                "x86_64",
+                "aarch64"
+            ]
+        },
+        {
+            "test_name": "build-serde-gnu",
+            "command": "cargo build --release --features=serde",
+            "platform": [
+                "x86_64",
+                "aarch64",
+                "riscv64"
+            ]
+        },
+        {
+            "test_name": "build-serde-musl",
+            "command": "cargo build --release --features=serde --target {target_platform}-unknown-linux-musl",
+            "platform": [
+                "x86_64",
+                "aarch64"
+            ]
+        }
+    ]
+}
```
These don't yet show up in buildkite because the pipeline steps don't include `./rust-vmm-ci/.buildkite/autogenerate_pipeline.py -t .buildkite/custom-tests.json | buildkite-agent pipeline upload`. Let me temporarily add it, retrigger the pipeline here. If it works, I'll remove it again and only permanently add it once this is merged (as otherwise we'll break CI on all other PRs)

roypat replied:
Seems to all work: https://buildkite.com/rust-vmm/kvm-ioctls-ci/builds/1250#_ :tada:

roypat replied:
Commented out the steps in the pipeline definition again until after we merge this

roypat reviewed (dismissed):
I've set the DCO check to "pass" because it's failing on ages old commits from kvm-bindings that look fine to me

andreeaflorescu commented:
Typo on commit 1e72627c98a0197b17c2fc5a6311d0ab0002d2a6: "Prepare for `kvm-bindings` to be merged, prevent conflicks from happening.".

Here is a suggestion for rephrasing your commit message to make it more explicit:

> monorepo: Move kvm-ioctls related files into kvm-ioctls folder
>
> This change  is needed to be able to merge the `kvm-bindings` crate into this repository
> and prevent conflicts from happening.

andreeaflorescu commented on .cargo/config:
```diff
@@ -1,5 +0,0 @@
-# This workaround is needed because the linker is unable to find __addtf3,
```
Why is this not needed anymore?

TimePrinciple replied:
I will keep it then :)

andreeaflorescu commented on kvm-ioctls/coverage_config_aarch64.json:
```diff
@@ -1,5 +0,0 @@
-{
-  "coverage_score": 77.1,
```
[optional] I would prefer to have an independent coverage test for the kvm-ioctls. For this we would need to add a custom test that runs pytest from the directory I think. If it's too complicated, we can also add a separate issue to handle it.

TimePrinciple replied:
Noted, I will work it out :)

TimePrinciple replied:
I just recollected if we were to do coverage test per crate, we'd better update our CI to utilize codecov. I will try to do it first and come back.

How about we leave this as it is for now :)

andreeaflorescu replied:
The current CI supports running individual coverage tests, no need to switch to codecov for this reason. But maybe codecov has some other features that make it a better choice. In any case, we can just open an issue and address this later. No need to spend time on it now.

andreeaflorescu commented on Cargo.toml:
```diff
@@ -0,0 +1,9 @@
+    "kvm-ioctls",
+]
+
+[workspace.dependencies]
```
Why does this need to be defined as a workspace dependency?

TimePrinciple replied:
IMHO keeping common dependencies here could help us better manage the monorepo, and send the message that we want these dependencies stay in sync :)

andreeaflorescu replied:
Looking at the documentation of workspace, it seems like you can't define the dependencies as optional: https://doc.rust-lang.org/nightly/cargo/reference/workspaces.html#the-dependencies-table. But we do need it to be optional for kvm-bindings. How does that work?


In any case, I would propose this change to be done in a separate PR because it is not strictly required for this merge, and we should discuss it separately.

TimePrinciple replied:
> Looking at the documentation of workspace, it seems like you can't define the dependencies as optional: https://doc.rust-lang.org/nightly/cargo/reference/workspaces.html#the-dependencies-table. But we do need it to be optional for kvm-bindings. How does that work?
> 
> In any case, I would propose this change to be done in a separate PR because it is not strictly required for this merge, and we should discuss it separately.

Good point, I will drop it then

andreeaflorescu commented on Cargo.toml:
```diff
@@ -0,0 +1,9 @@
+[workspace]
```
[nit] I don't quite understand what you mean with the commit message. Could you update it?

I guess what you want to say is: setup the repository as a Rust workspace?
And that the members of the workspace are kvm-bindings and kvm-ioctls?

You also have a typo in "contralize"

TimePrinciple replied:
Exactly, I will update it

andreeaflorescu commented on README.md:
```diff
@@ -0,0 +1,91 @@
+
+This repository contains the `kvm-bindings` and `kvm-ioctls` crates.
+
+## kvm-ioctls
```
[blocking for the merge] If you do this, then the crates.io readme will be empty (unless you update the Cargo.toml of each crate to specify the path to the new readme).

The right approach is to have each crate with their own README.md file, and on top of that, also have a top level readme that links to the individual readmes of the individual crates.

kvm-bindings can still be used without kvm-ioctls, so it doesn't make sense to have the readmes together.

TimePrinciple replied:
This was my idea of merging two READMEs together, I was thinking in a "workspace" way and didn't consider we need to publish them. Sorry for the inconvenience @roypat :(

andreeaflorescu replied:
I would propose to leave the readmes as they are in this PR, and just add a top-level readme. Then you can open a separate PR in which you update the kvm-bindings readme to fix the problems you've identified.

roypat replied:
Argh, sorry, this was my bad. Forgot about the crates.io thing. Will post a follow up!

RuoqingHe committed:
Merge kvm-bindings into kvm-ioctls

As discussed in #291, we are merging `kvm-bindings` into `kvm-ioctls`
and renaming `kvm-ioctls` to `kvm`.

This merges `kvm-bindings` into this workspace with its Git history
intact, through command:

```sh
git remote add bindings https://github.com/rust-vmm/kvm-bindings.git
git fetch bindings
git merge --allow-unrelated-histories bindings/main
```

Co-authored-by: Manos Pitsidianakis <manos.pitsidianakis@linaro.org>
Signed-off-by: Manos Pitsidianakis <manos.pitsidianakis@linaro.org>
Signed-off-by: Ruoqing He <heruoqing@iscas.ac.cn>

RuoqingHe committed:
chore: Update code coverage on x86_64

After merging `kvm-bindings` and `kvm-ioctls`, the code coverage has
changed, update coverage of `coverage_config_x86_64.json` to 91.46 as
suggested by coverage test.

Signed-off-by: Ruoqing He <heruoqing@iscas.ac.cn>

andreeaflorescu commented:
You're missing the top level README.md file. Are you planning to add it in a future PR?

TimePrinciple commented:
> You're missing the top level README.md file. Are you planning to add it in a future PR?

Yes, @roypat picked that up and I think it could be better to let him continue his work :)

roypat commented:
Similarly to https://github.com/rust-vmm/vhost/pull/114#issuecomment-1280773845, this PR cannot be merged with a rebase. I've also temporarily allowed merge commits to merge this

roypat commented:
I've
- updated the buildkite steps to include the custom-tests
- Updated the repository name
- Updated the buildkite pipeline name
- Updated the buildkite pipeline to fetch from the renamed repository

roypat commented:
Also updated branch protection rules to point towards the new pipeline name
